packages/frontend/ci/configFromEnvironment.js

```diff
@@ -0,0 +1,73 @@
+const assert = require("assert");
+
+const Environments = require("../../../features/environments.json");
+const { parseBooleanFromShell } = require("../src/config/envParsers");
+
+const Config = {
+  NEAR_WALLET_ENV: process.env.NEAR_WALLET_ENV,
+  TRAVIS: parseBooleanFromShell(process.env.TRAVIS),
+  BRANCH: process.env.BRANCH,
+  CLOUDFLARE_BASE_URL: process.env.CLOUDFLARE_BASE_URL,
+  CONTEXT: process.env.CONTEXT,
+  DEBUG_BUILD: process.env.DEBUG_BUILD,
+  DEPLOY_PRIME_URL: process.env.DEPLOY_PRIME_URL,
+  IS_DEVELOPMENT: process.env.NODE_ENV === "development",
+  IS_NETLIFY: parseBooleanFromShell(process.env.NETLIFY),
+  IS_RENDER: parseBooleanFromShell(process.env.RENDER),
+  IS_PULL_REQUEST: parseBooleanFromShell(process.env.IS_PULL_REQUEST),
+  RENDER: process.env.RENDER,
+  RENDER_EXTERNAL_URL: process.env.RENDER_EXTERNAL_URL,
+  RENDER_GIT_COMMIT: process.env.RENDER_GIT_COMMIT,
+  REVIEW_ID: process.env.REVIEW_ID,
+  SENTRY_AUTH_TOKEN: process.env.SENTRY_AUTH_TOKEN,
+  SENTRY_DSN: process.env.SENTRY_DSN,
+  SENTRY_RELEASE:
+      process.env.SENTRY_RELEASE ||
+      (process.env.RENDER &&
+          `render:${process.env.RENDER_SERVICE_NAME}:${process.env.RENDER_GIT_BRANCH}:${process.env.RENDER_GIT_COMMIT}`),
+  SHOULD_USE_CLOUDFLARE: parseBooleanFromShell(process.env.USE_CLOUDFLARE),
+};
+
+const computeCiNearWalletEnv = (Config) => {
+  if(Config.IS_DEVELOPMENT || Config.SHOULD_USE_CLOUDFLARE) {
+    return Environments.DEVELOPMENT;
+  }
+
+  if(Config.IS_NETLIFY) {
+    switch (Config.CONTEXT) {
+        case "production":
+            return Config.DEPLOY_PRIME_URL.includes("near-wallet-staging")
+                ? Environments.MAINNET_STAGING
+                : Environments.MAINNET;
+        case "branch-deploy":
+            return Environments.MAINNET_STAGING;
+        case "deploy-preview":
+            return Config.DEPLOY_PRIME_URL.includes("near-wallet-staging")
+                ? Environments.MAINNET_STAGING
+                : Environments.MAINNET;
+    }
+  }
+
+  if(Config.IS_RENDER) {
+    return Config.IS_PULL_REQUEST
+        ? Environments.TESTNET_STAGING
+        : Environments.TESTNET;
+  }
+
+  if(Config.TRAVIS) {
+    return Environments.MAINNET_STAGING;
+  }
+};
+
+const NEAR_WALLET_ENV =
+    Config.NEAR_WALLET_ENV || computeCiNearWalletEnv(Config);
```

Comment:
If `NEAR_WALLET_ENV` is not set as an environment variable but `computeCiNearWalletEnv` returns a valid value, the check in `src/config/configFromEnvironment.js` would still fail its assertion at runtime. The check here at bundle time enables the bundle to be built, but nothing has injected the value into the bundle, so any bundle code trying to parse `process.env.NEAR_WALLET_ENV` would get `undefined`.

If we can define (in code) the value for this environment variable then I think that addresses the issue. However I don't know whether that's possible for all environments:
- `development`: ✅  developers set this individually, either globally, at bundle time, or via the `.env` file (which implicitly uses `dotenv` to inject it as an environment variable)
- `testnet`: ✅  already set in the render.com configuration (not in code 😞)
- `testnet_STAGING`: ❌  the preview environment is using `testnet` because it uses the same render.com configuration - if there's no way to override this on the render.com side then I think the only way is to use the `IS_PULL_REQUEST` in conjunction with `NEAR_WALLET_ENV` to know that it's actually `testnet_STAGING`
- `mainnet_STAGING` ❌ (not yet available) it looks like we can set this one in `netlify.toml` under `[build.environment]` but I'm not sure
- `mainnet` ❌ (not yet available) same as `mainnet_STAGING` but set under `[context.production.environment]`

So I think we can address the `mainnet*` cases by adding the two lines in `netlify.toml` but I'm not sure what the ideal approach for `testnet_STAGING` would be, since it might be the only special case where we need to derive the value of `NEAR_WALLET_ENV` based on some other context.

Reply:
Thanks for catching this. I looked into this and I think the best approach is to write to the .env in case a value for `NEAR_WALLET_ENV` is not already set. Computing `NEAR_WALLET_ENV` separately for ci and bundle scatters the logic imo and logic is further scattered for the bundle's logic for each env. Let me know if you see any issues with this approach.

Reply:
This configuration will be less scattered as we transition to consistent environments in AWS with a monorepo housing the backend code, where we'll want the environment variables to be in the environment context at runtime. My preference would be to continue treating `.env` as a development artifact to be phased out, in particular since it's not a source-controlled file and would need to vary per environment anyway.

Looking at the CI environment variables more closely I see a lot of options that will be either deprecated or moved once we have a more robust build pipeline. I'm not sure exactly what this will look like but since it's more a concern for the environment context I'm thinking it will largely be a Terraform responsibility and not the responsibility of our app to determine. So while it is janky in its current state I think it's worth looking at what it would take to move the source of truth entirely to the environment. Keeping it an environmental concern will make the new pipeline simpler to transition to and reason about.

I dug into our Netlify configuration and saw that `NEAR_WALLET_ENV` is already set there (my bad, I should have checked this before 🤦), so actually `netlify.toml` doesn't need to be touched at all. The only real deviation is with `testnet_STAGING`, which I think this can be worked around with something like this (in `ci/configFromEnvironment`):
```
let { NEAR_WALLET_ENV } = process.env;
if (NEAR_WALLET_ENV = environments.TESTNET && Config.IS_PULL_REQUEST) {
    NEAR_WALLET_ENV = environments.TESTNET_STAGING;
}
```

Then in `src/config/configFromEnvironment` we can import `NEAR_WALLET_ENV` from the CI config or just duplicate the snippet above. Once it's set in both the build and runtime environments we'll be able to trust the environment and simply reference `process.env.NEAR_WALLET_ENV`.

What would you think about an approach more like that? Any potential problems you can see?

Reply:
I'd just like to avoid a scenario where we compute a different value of `NEAR_WALLET_ENV` in the ci from the bundle config and I think it's surprising that parcel does not allow the bundling script to define bundle envs directly in some way. The separate `NEAR_WALLET_ENV` computation when not explicitly provided will remain regardless of the stack because of this. It might be cleanest to import the value like you said. Importing was not working on my system because the bundle only gets env values from `.env` and not the parent shell for me so the values I set for testing in shell (e.g. `TRAVIS`) were not accessed by the bundle and `NEAR_WALLET_ENV` was still `undefined`. I'm not sure if this is the case in aws / render / netlify, it might be that the values we set in their UI are set as `.env` file and not just live shell. In the end we have to assume that whatever env values are accessible to the bundling script are accessible within the bundle and just import. The issue in my local just confused me a bit 😆 . Also I think precedence goes to `process.env` in the example above. We're dealing strictly with cases where `NEAR_WALLET_ENV` is not provided by the shell / environment.

Reply:
Facing an issue with [this](https://github.com/near/near-wallet/blob/master/features/features.js#L7) when we don't set the value to `process.env` in the bundle and just import. It takes whatever value is in `.env` and throws if it's not set. This issue is not there when we write to `.env` from bundling script like I did, but if we go this route I think we need to use `dotenv` on the bundling script as well. Assuming that we don't want to import `Config` in that directory and also that we can't set `process.env` to the imported value from within the bundle, this seems to be the only way to set `process.env` directly while maintaining consistency.

Reply:
> We're dealing strictly with cases where `NEAR_WALLET_ENV` is not provided by the shell / environment.
My point is more that we _can_ ensure that it is provided by the environment, and we currently do by configuring it on the various hosting providers. It's not perfect in its current iteration because of `testnet_STAGING`, but that's an implementation detail in Render that we can work around 🙂. It will be more reliable when we are under a single provider. Moreover since it is an environmental concern any computation used to derive it independently gives us two sources of truth, which we definitely want to avoid.

> Importing was not working on my system because the bundle only gets env values from `.env` and not the parent shell for me so the values I set for testing in shell (e.g. `TRAVIS`) were not accessed by the bundle and `NEAR_WALLET_ENV` was still `undefined`.
Since the command we run to bundle, `node ci/runBundler.js`, is executed in a child process, it only inherits environment variables, not shell variables. This wasn't a distinction I was aware of before but your comment led me to experiment a bit. Basically if you are setting (in your `.zshrc` or wherever) `NEAR_WALLET_ENV=development` then it is a shell variable but if you instead have `export NEAR_WALLET_ENV=development` it is an environment variable. Only the latter will be available to the child process context.

`dotenv` is only ever a concern in our local environments because no `.env` files exist in production environments. The bundling is always done using the latest checkout and so must rely exclusively on the environment variables set at bundle time.

Reply:
Made it so that `src/config/configFromEnvironment` and `features/features.js` both import directly from `ci/configFromEnvironment.js`. This does not set `process.env.NEAR_WALLET_ENV` within the bundle as far as parcel is concerned but all references to `NEAR_WALLET_ENV` will use whatever was computed in `ci/configFromEnvironment.js` since they point to `src/config/configFromEnvironment`.

In cases where `process.env.NEAR_WALLET_ENV` is provided, that will be directly used and in cases where it's not the logic on `computeCiNearWalletEnv` should compute it correctly since it's based on `composeBundlerConfig`  in `ParcelBundler.js`.

e.g `NEAR_WALLET_ENV` will be `testnet_STAGING` when `IS_RENDER` and `IS_PULL_REQUEST` are both true. Let me know if you see something I'm missing 😄

Reply:
As discussed, removing the fallback computation also allows us to directly access `NEAR_WALLET_ENV` from `process.env` in bundle and `features.js` so I removed the additional wiring we decided to add earlier as well 👍